botorch/optim/closures.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+r"""Model fitting closures"""
+
+from __future__ import annotations
+
+from itertools import islice
+from typing import Any, Callable, Optional, Sequence
+
+from botorch.utils.dispatcher import Dispatcher
+from gpytorch.mlls import (
+    ExactMarginalLogLikelihood,
+    MarginalLogLikelihood,
+    SumMarginalLogLikelihood,
+)
+from torch import Tensor
+from torch.utils.data import DataLoader
+
+dispatcher = Dispatcher("get_loss_closure")
+NoneType = type(None)
+TLossClosure = Callable[[], Tensor]
+
+
+def get_loss_closure(
+    mll: MarginalLogLikelihood, data_loader: Optional[DataLoader] = None, **kwargs: Any
+) -> Callable[[], Tensor]:
+    return dispatcher(data_loader, mll)
+
+
+@dispatcher.register(DataLoader, MarginalLogLikelihood)
+def _get_closure_external_fallback(
+    data_loader: DataLoader, mll: MarginalLogLikelihood, **ignore: Any
+):
+    r"""Fallback closure for external training data."""
+
+    def _closure(**kwargs: Any) -> Tensor:
+        tensors = next(iter(data_loader))
+        if not isinstance(tensors, Sequence):
+            raise TypeError
+
+        tensors_iter = iter(tensors)
+        model_output = mll.model(*islice(tensors_iter, len(mll.model.train_inputs)))
```

Comment:
So this will ignore all excess inputs provided by `tensors_iter` that are at a position later than the length of the training inputs. Is this what we want here or should we error out if these are not compatible in size? If we silently drop those this could be hard to debug down the road.

Reply:
They aren't dropped actually. We expand `tensors_iter` when calling `mll` below and this consumes the remainder of the iterator.